ntoskrnl/io/pnpmgr/pnpmgr.c

```diff
@@ -1760,6 +1765,80 @@ IopGetParentIdPrefix(PDEVICE_NODE DeviceNode,
     return Status;
 }
 
+static
+BOOLEAN
+IopValidateID(
+    _In_ PWCHAR Id,
+    _In_ SIZE_T MaxIdLen,
+    _In_ ULONG MaxSeparators,
+    _In_ BOOLEAN IsMultiSz)
+{
+    PWCHAR PtrSymbol;
+    PWCHAR StringEnd;
+    WCHAR Symbol;
+    ULONG SeparatorsCount = 0;
+    PWCHAR PtrPrevSymbol = NULL;
+
+    PAGED_CODE();
+
+    StringEnd = Id + MaxIdLen;
+
+    for (PtrSymbol = Id; PtrSymbol < StringEnd; PtrSymbol++)
```

Comment:
This seems like it would be simpler with a "for i" loop rather than "for ptr"

Reply:
```
    for (ix = 1; ix <= MAX_DEVICE_ID_LEN; ix++)
    {
        Char = *PtrChar;

        if (Char == UNICODE_NULL)
        {
            if (!IsMultiSz || (PtrPrevChar && PtrChar == PtrPrevChar + 1))
            {
            }

            ix = 1;
            PtrPrevChar = PtrChar;
        }
        else if ()
        {
        }
        else if ()
        {
        }
        else if ()
        {
        }

        PtrChar++; 
    }
```
Or?

Reply:
I would make i start at zero instead of one, and replace PtrChar by Id[i] as well as make PtrPrevChar an index instead of a pointer.

Reply:
If so
```
    for (ix = 0, jx = 0; ix < MAX_DEVICE_ID_LEN; ix++)
    {
        Char = Id[ix + jx];

        if (Char == UNICODE_NULL)
        {
            if (!IsMultiSz ||
                (Id[ix + jx - 1] && Id[ix + jx] == Id[ix + jx - 1]))
            {
            }

            jx = ix + 1;
            ix = 0;
        }
        else if (Char < ' ' || Char > 0x7F || Char == ',')
        {
        }
        else if (Char == ' ')
        {
            Id[ix + jx] = '_';
        }
        else if (Char == '\\')
        {
        }
    }
```
then I again have a question: what will happen in the case of an empty string (ID [0] == NULL and ID [1] == NULL)?